### PR TITLE
More TIS features; skip recompute; mismatch metrics without TIS

### PR DESCRIPTION
**Pull request by yueming-yuan**
### reference
https://github.com/szrlee/verl/blob/yingru/rollout_correction/docs/advance/rollout_corr.md

https://github.com/szrlee/verl/blob/yingru/rollout_correction/docs/advance/rollout_corr_math.md

### TODO:
<ul>
<li>Test more cases.</li>
<li>Add detailed discussion on IS mode/level choice.</li>
<li>Add documentation on arguments usage.</li>
<li>Add more details (formulas/pseudo codes) in this PR's description.</li>
</ul>

### changes

This PR contains the new algorithms on training-inference importance sampling. 
Major modification:
1. Skip `actor.compute_log_probs()` when enabling `args.use_rollout_logprobs`. This mode is faster and saves the computation of reforward pass.
2. Follow the algorithms mentioned in [rollout correction math](https://github.com/szrlee/verl/blob/yingru/rollout_correction/docs/advance/rollout_corr_math.md), implemented the REINFORCE + pure IS algorithm.
3. Aligned the APIs and algorithms, as shown in the table below.

**Functionality before and after this change：**

args.use_rollout_logprobs | args.use_tis | behavior before | behavior after
-- | -- | -- | --
False | False | Standard PPO | (no change)
False | True | Decoupled PPO, 3-policies with TIS/MIS; use old_log_probs to calculate | (no change)
True | False | Overwrite PPO’s IS with pi_theta/pi_rollout; **still do re-forward and use old_log_probs to calculate KL inside PPO advantages** | Overwrite PPO’s IS with pi_theta/pi_rollout; **skip the re-forward pass and directly use rollout_log_probs to calculate PPO advantages**
True | True | **Not allowed**; asserted to not be set simultaneously | Use pure_is algorithm. **skip the re-forward pass. skip the PPO clipping. Directly apply REINFORCE loss + rollout correction**

**Algorithms explanation for each mode [in construction, will add more details soon]:**
<ol>
<li>
<p>Standard PPO <code>args.use_rollout_logprobs=False</code> + <code>args.use_tis=False</code></p>
<p>Standard PPO, without training-inference correction.</p>
<p>actor.compute_log_probs(): required.</p>
</li>
<li>
<p>Bypass standard PPO <code>args.use_rollout_logprobs=True</code> + <code>args.use_tis=False</code></p>
<p>Correct training-inference error by bypassing</p>
<p>actor.compute_log_probs(): not required. Use rollout_log_probs instead.</p>
</li>
<li>
<p>Decoupled PPO <code>args.use_rollout_logprobs=False</code> + <code>args.use_tis=True</code></p>
<p>Correct training-inference error by bypassing</p>
<p>actor.compute_log_probs(): required.</p>
</li>
<li>
<p> REINFORCE + Pure IS<code>args.use_rollout_logprobs=True</code> + <code>args.use_tis=True</code></p>
<p> REINFORCE + Pure IS. without PPO clipping. </p> 
<p>actor.compute_log_probs(): not required. Use rollout_log_probs instead.</p>
</li>
</ol>

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
Great work! Sorry for the delay of review. I will go over:

https://github.com/szrlee/verl/blob/yingru/rollout_correction/docs/advance/rollout_corr.md

https://github.com/szrlee/verl/blob/yingru/rollout_correction/docs/advance/rollout_corr_math.md

Then continue to review. In the same time, feel free to push me for the review.

**Comment by zhaochenyang20** [comment]
https://github.com/yueming-yuan/slime/blob/rollout_correction/docs/en/advanced/rollout-correction.md

Read this also

**Comment by ChangyiYang** [comment]
Let's goooo!

**Comment by ChangyiYang** [comment]
I think TIS can also be used when using GRPO? Maybe can add some explaination the behavior change when using GRPO.

File: slime/backends/megatron_utils/loss.py [hunk 1]
```diff
@@ -361,6 +366,126 @@ def compute_advantages_and_returns(args: Namespace, rollout_data: RolloutBatch)
     rollout_data["returns"] = returns
 
 
+def get_rollout_training_metrics(
```
Comment (guapisolo) [inline slime/backends/megatron_utils/loss.py]: Move to `examples/train_infer_mismatch_helper/mis.py`.

File: slime/backends/megatron_utils/loss.py [hunk 2]
```diff
@@ -361,6 +366,126 @@ def compute_advantages_and_returns(args: Namespace, rollout_data: RolloutBatch)
...
+        "mismatch_k3_kl": k3_kl_per_token,
+        "mismatch_log_ppl_diff": log_ppl_diff,
+        "mismatch_log_ppl_abs_diff": log_ppl_diff.abs(),
+        "mismatch_ppl_ratio": ppl_ratio,
+        "mismatch_chi2_token": chi2_token,
+        "mismatch_chi2_seq": chi2_seq,
+    }
+
+
+def _compute_metrics_with_cp(
```
Comment (guapisolo) [inline slime/backends/megatron_utils/loss.py]: Move to `examples/train_infer_mismatch_helper/mis.py`.

File: slime/backends/megatron_utils/loss.py [hunk 3]
```diff
@@ -408,34 +533,42 @@ def policy_loss_function(
...
-        loss_masks = batch["loss_masks"]
-        ppo_kl = [
-            ((old_logprob - log_prob) * loss_mask).sum() / torch.clamp_min(loss_mask.sum(), 1)
-            for log_prob, old_logprob, loss_mask in zip(full_log_probs, full_old_log_probs, loss_masks)
-        ]
-        ppo_kl = [kl.expand_as(log_prob) for kl, log_prob in zip(ppo_kl, log_probs)]
-        ppo_kl = torch.cat(ppo_kl, dim=0)
-        old_log_probs = torch.cat(old_log_probs, dim=0)
+    if args.use_rollout_correction and args.use_rollout_logprobs:
+        # skipping clip, use pure reinforce + rollout correction
```
Comment (guapisolo) [inline slime/backends/megatron_utils/loss.py]: Move the comment to `examples/train_infer_mismatch_helper/mis.py`.

File: slime/utils/arguments.py [hunk 4]
```diff
@@ -748,6 +748,12 @@ def add_algo_arguments(parser):
                 ),
             )
             # Off-Policy Correction using Importance Sampling: https://fengyao.notion.site/off-policy-rl
+            parser.add_argument(
```
Comment (guapisolo) [inline slime/utils/arguments.py]: I think most of users only know tis and add this argument is a little bit confusing.
Does "--custom-tis-function-path" cannot support the current demand?

**Comment by zhaochenyang20** [comment]
Nice catch @guapisolo

**Review by zhuzilin — APPROVED** [review]
